### DISCUSSION TITLE
Lambda access to stackalloc variables

### DISCUSSION THREAD

**Discussion by Damien-The-Unbeliever** [dotnet/csharplang#2182] [Language Ideas]
Someone has posed a [question on Stackoverflow ](https://stackoverflow.com/q/54414317/15498) asking why a `stackalloc` variable isn't correctly captured by a lambda.

The more I think about this, the more I think that just writing such a thing should produce a warning - lambdas have an uncertain lifetime that may live longer than the stack frame in which they're created, but `stackalloc` is an explicit statement that the lifetime of an object should only last as long as the current stack frame. There doesn't seem to be a natural way for these two features to interact safely.

But as currently implemented, no error or warning is produced by this code:

```
using System;
using System.Threading;

class Program
{
    static void Main(string[] args)
    {
        for (int i = 0; i < 5; i++)
            call(i);

        Thread.Sleep(1000);
    }

    static unsafe void call(int number)
    {
        int* ints = stackalloc int[64];

        ints[32] = number;

        ThreadPool.QueueUserWorkItem(delegate { execute(ints[32]); });
    }

    static void execute(int number)
    {
        Console.WriteLine(" * NUM=" + number.ToString());
    }
}
```
(Copied from question)

Should at least a new warning be produced for such code? (Because it may be known by the author that the useful life of the lambda doesn't extend beyond the lifetime of the stackframe, I'm thinking an error wouldn't be appropriate)

**Comment by Damien-The-Unbeliever** [comment]
And now I'm starting to doubt myself because it is, after all, unsafe code. You told the compiler to trust you, the fact that you betrayed that trust isn't going to always be discoverable.

**Comment by YairHalberstadt** [comment]
I guess the issue with that is that unsafe code is explicitly that - unsafe.

Someone who has opted into unsafe code is meant to know about the risks, and deal with them. He doesn't want to be given warnings about dangerous code. If course it's dangerous! That's why it's unsafe!

Adding a warning to code that already existed and works fine will just annoy them.

**Comment by MkazemAkhgary** [comment]
I wish there was a chaos keyword too.

**Comment by AlgorithmsAreCool** [comment]
@Damien-The-Unbeliever 
For what it's worth if you use the new `Span<>` type instead of pointers [the compiler will stop you from making this mistake.](https://sharplab.io/#v2:C4LgTgrgdgNAJiA1AHwAIAYAEqCMBuAWACgNscA6AFQAswBTAQzgEsoBzQo41AJjIHZiAb2KYxZAGzYALJgCyDVgApc6ANoBdTAzBsAzgEpR4kUXHnMAMwD2YTEtbBMzTAF5M6PM8wAeTAFYvZkREIzMLCIBjBgAbGIcDTmMImnomcgBlGLo6AAclHHQixOTMAF9iUtwpVFlouIcoJygIAFsAIzowMPNTCLEM3IYoH0cAPmcmvTdMPWAGSIBrWJjrSMngNQlpDSTwiMc9NQBmHi13Fo6uvf7MVMY4AAVraxjyAEUIOi+AVT0ugDqtkWAElgHRWko4HRsmwGODMEJMHQAB50SIQcGNYBHU4aRLlEr7cqVYnVGTItEYrGOTCXTrdUp9fq4ACcSgARJgAFSYAByPzkri5iDpbQZVGsGWAYFYbCUBiJ5gqRDKQA=)


```csharp
using System;
using System.Threading;

class Program
{
    static void Main(string[] args)
    {
        for (int i = 0; i < 5; i++)
            call(i);

        Thread.Sleep(1000);
    }

    static void call(int number) //this is no longer 'unsafe'
    {
        Span<int> ints = stackalloc int[64];

        ints[32] = number;

        ThreadPool.QueueUserWorkItem(delegate { execute(ints[32]); }); //<-- error CS8175
    }

    static void execute(int number)
    {
        Console.WriteLine(" * NUM=" + number.ToString());
    }
}
```
